firestore/bulkwriter.go

```diff
@@ -0,0 +1,521 @@
+	MaxBatchSize = 20
+	// RetryMaxBatchSize is the max number of writes to send in a retry request
+	RetryMaxBatchSize = 10
+	// MaxRetryAttempts is the max number of times to retry a write
+	MaxRetryAttempts = 10
+	// DefaultStartingMaximumOpsPerSecond is the starting max number of requests to the service per second
+	DefaultStartingMaximumOpsPerSecond = 500
+	// RateLimiterMultiplier is the amount to increase maximum ops (qps) every 5 minutes
+	RateLimiterMultiplier       = 1.5
+	rateLimiterMultiplierMillis = 5 * 60 * 1000 // 5 minutes in milliseconds
+	coolingPeriodMillis         = 2.0           // starting time to wait in between requests
+)
+
+type bulkWriterJob struct {
+	err      chan error           // send error responses to this channel
+	result   chan *pb.WriteResult // send the write results to this channel
+	write    *pb.Write            // the writes to apply to the database
+	attempts int                  // number of times this write has been attempted
+}
+
+type bulkWriterBatch struct {
+	bwr *pb.BatchWriteRequest // request to send to the service
+	bwj []bulkWriterJob       // corresponding jobs to return response
+}
+
+type bulkWriterRequestBatch struct {
+	isFlushing bool              // flag that indicates that BulkWriter needs to schedule a flush of the queue
+	bwb        []bulkWriterBatch // all the bulkWriterBatch objects to schedule
+}
+
+type bulkWriterStats int
+
+const (
+	writesProvided bulkWriterStats = iota
+	writesSent
+	writesReceived
+)
+
+// The BulkWriterStatus provides details about the BulkWriter, including the
+// number of writes processed, the number of requests sent to the service, and
+// the number of writes in the queue.
+type BulkWriterStatus struct {
+	WritesProvidedCount int  // number of write requests provided by caller
+	IsOpen              bool // whether this BulkWriter is open or closed
+	WritesSentCount     int  // number of requests sent to the service
+	WritesReceivedCount int  // number of WriteResults received from the service
+}
+
+// A BulkWriter allows multiple document writes in parallel. The BulkWriter
+// submits document writes in maximum batches of 20 writes per request. Each
+// request can contain many different document writes: create, delete, update,
+// and set are all supported. Only one operation per document is allowed.
+// Each call to Create, Update, Set, and Delete can return a value and error.
+type BulkWriter struct {
+	database        string                      // the database as resource name: projects/[PROJECT]/databases/[DATABASE]
+	reqs            int64                       // total number of requests sent
+	start           time.Time                   // when this BulkWriter was started
+	vc              *vkit.Client                // internal client
+	isOpen          bool                        // signal that the BulkWriter is closed
+	startFlush      chan struct{}               // signal the beginning of flushing operation
```

Comment:
There seems to be quite a bit of sync complexity in this PR. Did you consider or look at using the bundler package to encapsalte this complexity? I think it may support many of the features you may need in this PR. We have examples in other libs how it is used internally: https://pkg.go.dev/google.golang.org/api/support/bundler

Reply:
I only looked at it cursorily (apologies--I know you sent me in that direction!). I thought it would be good to have an initial version to start the conversation with.